### PR TITLE
CB-8206: Browser platform: Add support for update

### PR DESCRIPTION
**Pull request by nikhilkh**
This change add supports for updating the browser platform. I ended up deleting a couple of log messages from create as there is no way to retrieve and log them on update.

### CODE REVIEW

File: bin/lib/update.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        shell.config.fatal = true;
+        var returnVal = shellFunc.apply(shell, slicedArgs);
+    }   
+    finally
+    {
+        shell.config.fatal = false;
+    }
+    return returnVal;
+}
```
Comment (jsoref) [inline bin/lib/update.js]: The red mark here means that you didn't insert a newline at the end of your file. please do insert one.

File: bin/lib/create.js [hunk 2]
```diff
@@ -26,14 +26,11 @@ var fs = require('fs'),
     ROOT    = path.join(__dirname, '..', '..'),
     check_reqs = require('./check_reqs');
 
-
 exports.createProject = function(project_path,package_name,project_name){
     var VERSION = fs.readFileSync(path.join(ROOT, 'VERSION'), 'utf-8');
     
     // Set default values for path, package and name
     project_path = typeof project_path !== 'undefined' ? project_path : "CordovaExample";
-    package_name = typeof package_name !== 'undefined' ? package_name : 'org.apache.cordova.example';
```
Comment (jsoref) [inline bin/lib/create.js]: this looks suspiciously like you saved after updating and overwrote someone else's changes.

in general, adding a feature (like `update`) shouldn't have any impact to a parallel command (like `create`)

File: node_modules/.bin/shjs [hunk 3]
```diff
@@ -1 +1,15 @@
-../shelljs/bin/shjs
```
Comment (jsoref) [inline node_modules/.bin/shjs]: this change looks like you replaced a *nix symlink w/ a file.

the change shouldn't be done in the same commit as adding support for `update`.
and you should probably do something to ensure that someone doesn't later undo this when working from *nix.

File: bin/lib/create.js (hunk 2 above)
Comment (nikhilkh) [inline bin/lib/create.js]: Turns out that `update` depends on `create`. As I mentioned in my pull request, I had to remove these console.log messages as for the browser platform we do not save the package and project name on creation. On upgrade, I do not have a package and project name when I invoke `createProject`. These messages had marginal value so I decided to remove them.

File: bin/lib/update.js (hunk 1 above)
Comment (nikhilkh) [inline bin/lib/update.js]: Interesting! I will fix this. I'm curious - what's the rationale behind this? Is this just a styling best practice on this project or in JavaScript community?

Comment (purplecabbage) [inline bin/lib/update.js]: http://stackoverflow.com/questions/729692/why-should-files-end-with-a-newline
Generally a best practice, not just js, and not just this project.

File: node_modules/.bin/shjs (hunk 3 above)
Comment (nikhilkh) [inline node_modules/.bin/shjs]: This is done automagically by node on windows to create a shim for a *nix batch file: http://stackoverflow.com/questions/10396305/npm-package-bin-script-for-windows. While I agree with you that this is not strictly related to `update` change, since this is in the `node_modules` directory, I would prefer to keep this as is. Let me know if you still disagree.

Comment (mmocny) [inline node_modules/.bin/shjs]: Why are node_modules even checked in?  Shouldn't they just be added to .gitignore?  I double checked, and the only other cordova repo that checks these in is firefoxos, which I think is also an oversight.

Comment (jsoref) [inline node_modules/.bin/shjs]: @mmocny : checking them in is the only way things actually work, otherwise offline installs and various other things break.

blackberry10 checks in its node_modules too

Comment (jsoref) [inline node_modules/.bin/shjs]: @nikhilkh : i would prefer you split it into a distinct commit (before this commit) -- see http://stackoverflow.com/questions/4307095/git-how-to-split-up-a-commit-buried-in-history for how to do it.

Comment (mmocny) [inline node_modules/.bin/shjs]: For offline installs you still need to pre-fetch, where you can do the npm install step as well.

And I'm not at this point discussing what does into an RC, just what's stored in our git master.

**Comment by jsoref** [comment]
@nikhilkh: please use `git rebase` to fold the fixes for 2d77f2b into aad4c25

**Comment by jsoref** [comment]
well, err, specifically: for lines you're adding in aad4c25 with bad whitespace, you should fold in the fixes from 2d77f2b. If you're fixing the entire file, you can do that as a distinct commit.

File: node_modules/.bin/shjs (hunk 3 above)
Comment (nikhilkh) [inline node_modules/.bin/shjs]: Good question @mmocny and you're not the first one to ask that: http://callback.markmail.org/thread/vl35coc24d4vjdph. Cordova-windows also does this. I tend to agree with you that this is less than ideal design. As mentioned in the thread, there is a couple of technical limitations currently and @jsoref expressed some other concerns.

**Comment by nikhilkh** [comment]
@jsoref Thanks for reviewing this and the instructions for doing the git magic. It was useful. I have made the changes that you requested.

**Comment by nikhilkh** [comment]
Looking for anyone who can review this and commit it? It's quite a simple change.

File: bin/lib/create.js [hunk 4]
```diff
@@ -75,7 +69,7 @@ exports.createProject = function(project_path,package_name,project_name){
         'run',
         'build',
         'clean',
-        'version',
+        'version'
```
Comment (jsoref) [inline bin/lib/create.js]: you shouldn't make this change. trailing commas are legal and in fact encouraged for arrays, as they make diffs easier to read.

File: bin/update [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+if (['--help', '/?', '-h', 'help', '-help', '/help'].indexOf(process.argv[2]) > -1) {
+    update.help();
+} else {
+    update.run(process.argv).done(function () {
+        console.log('Successfully updated browser project.');
+    }, function (err) {
+        console.error('Update failed due to', err);
+        process.exit(2);
+    });
+}
```
Comment (jsoref) [inline bin/update]: please ensure that there's a new line at the end of each file, when you see a red null/carriage mark w/ a green background, that means that the diff system is pointing out that you're missing the newline.

**Comment by jsoref** [comment]
I'd like someone else to review this -- but please address my other comments. Sorry for the multipass review.

And, I need to see about ensuring there's a checklist (on the wiki or in coho or something) for most of these points...

File: bin/lib/create.js (hunk 4 above)
Comment (nikhilkh) [inline bin/lib/create.js]: Fixed.

File: bin/update (hunk 5 above)
Comment (nikhilkh) [inline bin/update]: Sorry that I did not update all the files correctly earlier. It would be great to automate something like this. :)

**Comment by nikhilkh** [comment]
@jsoref Your comments have been addressed. 

 I will wait for someone else to review this. A checklist would be great - looks like there is a doc here: https://github.com/apache/cordova-coho/blob/master/docs/code-reviews.md but lacks details.

**Comment by mmocny** [comment]
I reviewed this in a previous iteration, and the latest changes look fine.

Since this feature does not exist yet anyway, and I'm not too concerned about inline updates to browser platform, I think this is more than ready to ship.

**Comment by mmocny** [comment]
..Thanks!

**Comment by mmocny** [comment]
(Josh I'll let you have the last word on landing, you have been running this)

**Comment by jsoref** [comment]
ok, one last final thing:
* Please amend the commit message to not have the `:` after the `CB-8206` (I'm fairly confident this is documented in our instructions).

Then, someone can push this to cordova-browser.

**Comment by nikhilkh** [comment]
@jsoref Updated the commit message. Please review and help merge to master.
